java.security.PrivilegedActionException: java.security.PrivilegedActionException: java.net.SocketTimeoutException: Read timed out

### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
PyCharm 2022.1.1 (build PY-221.5591.52)

### Operating System 
Windows 10 10.0 (amd64)

### TeXiFy IDEA version
0.7.18

### Description



### Stacktrace
```
java.security.PrivilegedActionException: java.security.PrivilegedActionException: java.net.SocketTimeoutException: Read timed out
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:898)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:743)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:439)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:803)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:438)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:119)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:604)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:436)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:873)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:484)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.security.PrivilegedActionException: java.net.SocketTimeoutException: Read timed out
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	... 20 more
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.base/java.net.SocketInputStream.socketRead0(Native Method)
	at java.base/java.net.SocketInputStream.socketRead(SocketInputStream.java:115)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:168)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)
	at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:478)
	at java.base/sun.security.ssl.SSLSocketInputRecord.readFully(SSLSocketInputRecord.java:461)
	at java.base/sun.security.ssl.SSLSocketInputRecord.decodeInputRecord(SSLSocketInputRecord.java:243)
	at java.base/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:181)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:111)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1504)
	at java.base/sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1471)
	at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1060)
	at java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:290)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:351)
	at java.base/sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:555)
	at java.base/sun.net.www.http.ChunkedInputStream.readAhead(ChunkedInputStream.java:612)
	at java.base/sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:699)
	at java.base/java.io.FilterInputStream.read(FilterInputStream.java:133)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3514)
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.base/java.io.InputStreamReader.read(InputStreamReader.java:181)
	at java.base/java.io.Reader.read(Reader.java:229)
	at kotlin.io.TextStreamsKt.copyTo(ReadWrite.kt:127)
	at kotlin.io.TextStreamsKt.copyTo$default(ReadWrite.kt:120)
	at kotlin.io.TextStreamsKt.readText(ReadWrite.kt:107)
	at nl.hannahsten.texifyidea.LatexErrorReportSubmitter$Companion.getLatestVersion(LatexErrorReportSubmitter.kt:138)
	at nl.hannahsten.texifyidea.LatexErrorReportSubmitter.submit(LatexErrorReportSubmitter.kt:45)
	at com.intellij.diagnostic.IdeErrorsDialog.reportMessage(IdeErrorsDialog.java:635)
	at com.intellij.diagnostic.IdeErrorsDialog.reportAll(IdeErrorsDialog.java:1103)
	at com.intellij.diagnostic.IdeErrorsDialog$ReportAndClearAllAction.actionPerformed(IdeErrorsDialog.java:1087)
	at com.intellij.diagnostic.IdeErrorsDialog$CompositeAction.actionPerformed(IdeErrorsDialog.java:994)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:297)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)

```

--- a/src/nl/hannahsten/texifyidea/LatexErrorReportSubmitter.kt
+++ b/src/nl/hannahsten/texifyidea/LatexErrorReportSubmitter.kt
@@ -11,18 +11,14 @@ import com.intellij.openapi.diagnostic.ErrorReportSubmitter
 import com.intellij.openapi.diagnostic.IdeaLoggingEvent
 import com.intellij.openapi.diagnostic.SubmittedReportInfo
 import com.intellij.openapi.extensions.PluginId
-import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.ui.MessageDialogBuilder
 import com.intellij.openapi.ui.popup.JBPopupFactory
-import com.intellij.openapi.ui.showOkCancelDialog
 import com.intellij.openapi.util.SystemInfo
 import com.intellij.util.Consumer
 import org.apache.maven.artifact.versioning.DefaultArtifactVersion
 import java.awt.Component
 import java.io.UnsupportedEncodingException
-import java.net.HttpURLConnection
-import java.net.URL
-import java.net.URLEncoder
-import java.net.UnknownHostException
+import java.net.*
 
 /**
  * Send error report to GitHub issue tracker.
@@ -31,7 +27,6 @@ import java.net.UnknownHostException
  */
 class LatexErrorReportSubmitter : ErrorReportSubmitter() {
 
-    @Suppress("DialogTitleCapitalization")
     override fun getReportActionText() = "Report to TeXiFy-IDEA issue tracker"
 
     override fun submit(
@@ -55,16 +50,18 @@ class LatexErrorReportSubmitter : ErrorReportSubmitter() {
 
             JBPopupFactory.getInstance().createMessage("")
                 .showInCenterOf(parentComponent)
-            val result = showOkCancelDialog(
-                "Update TeXiFy",
-                "Please update your current version ($currentVersion) of TeXiFy to the latest version (${latestVersion.version}) before submitting,\n" +
-                    "to check if the error is already fixed. Go to Settings > Plugins to update.\n" +
-                    if (currentIdeaVersion < requiredIdeaVersion) "You first need to update your current IDE version ($currentIdeaVersion) to $requiredIdeaVersion or newer.\n" else "",
-                "Cancel Submit", // Sort of the wrong way around, but it suggests to cancel this way
-                "Submit Anyway"
-            )
 
-            if (result == Messages.OK) return false
+            val message = "Please update your current version ($currentVersion) of TeXiFy to the latest version (${latestVersion.version}) before submitting,\n" +
+                    "to check if the error is already fixed. Go to Settings > Plugins to update.\n" +
+                    if (currentIdeaVersion < requiredIdeaVersion) "You first need to update your current IDE version ($currentIdeaVersion) to $requiredIdeaVersion or newer.\n" else ""
+
+            val result = MessageDialogBuilder.okCancel("Update TeXiFy", message)
+                .yesText("Cancel Submit") // Sort of the wrong way around, but it suggests to cancel this way
+                .noText("Submit Anyway")
+                .ask(parentComponent)
+
+            // If the user cancels, don't submit
+            return !result
         }
 
         submit(events, additionalInfo, consumer)
@@ -135,7 +132,12 @@ class LatexErrorReportSubmitter : ErrorReportSubmitter() {
                 connectTimeout = 1000
                 readTimeout = 1000
 
-                val inputString: String = inputStream.reader().use { it.readText() }
+                val inputString: String = try {
+                    inputStream.reader().use { it.readText() }
+                }
+                catch (e: SocketTimeoutException) {
+                    return latestVersionCached
+                }
                 latestVersionCached = mapper.readValue(inputString, PluginRepo::class.java)
                     .category
                     ?.maxByOrNull { it.version }

--- a/src/nl/hannahsten/texifyidea/editor/typedhandlers/InlineMathBackspaceHandler.kt
+++ b/src/nl/hannahsten/texifyidea/editor/typedhandlers/InlineMathBackspaceHandler.kt
@@ -1,7 +1,6 @@
 package nl.hannahsten.texifyidea.editor.typedhandlers
 
 import com.intellij.codeInsight.editorActions.BackspaceHandlerDelegate
-import com.intellij.grazie.utils.orFalse
 import com.intellij.openapi.editor.Editor
 import com.intellij.psi.PsiFile
 import com.intellij.psi.util.PsiUtilBase.getElementAtCaret
@@ -11,6 +10,7 @@ import nl.hannahsten.texifyidea.util.files.document
 import nl.hannahsten.texifyidea.util.files.isLatexFile
 import nl.hannahsten.texifyidea.util.get
 import nl.hannahsten.texifyidea.util.inVerbatim
+import nl.hannahsten.texifyidea.util.orFalse
 
 /**
  * In the situation $<caret>$, pressing backspace should delete both $.

--- a/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexTypedHandler.kt
+++ b/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexTypedHandler.kt
@@ -2,7 +2,6 @@ package nl.hannahsten.texifyidea.editor.typedhandlers
 
 import com.intellij.codeInsight.CodeInsightSettings
 import com.intellij.codeInsight.editorActions.TypedHandlerDelegate
-import com.intellij.grazie.utils.orFalse
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.editor.ex.EditorEx
 import com.intellij.openapi.fileTypes.FileType
@@ -19,6 +18,7 @@ import nl.hannahsten.texifyidea.psi.LatexTypes
 import nl.hannahsten.texifyidea.settings.TexifySettings.Companion.getInstance
 import nl.hannahsten.texifyidea.util.files.isLatexFile
 import nl.hannahsten.texifyidea.util.inVerbatim
+import nl.hannahsten.texifyidea.util.orFalse
 
 /**
  * @author Sten Wessel

--- a/src/nl/hannahsten/texifyidea/settings/sdk/MiktexWindowsSdk.kt
+++ b/src/nl/hannahsten/texifyidea/settings/sdk/MiktexWindowsSdk.kt
@@ -36,8 +36,11 @@ class MiktexWindowsSdk : LatexSdk("MiKTeX Windows SDK") {
         val paths = "where pdflatex".runCommand()
         if (paths != null && !paths.contains("Could not find")) { // Full output is INFO: Could not find files for the given pattern(s).
             paths.split("\r\n").forEach { path ->
+                if (path.isBlank()) return@forEach
                 val index = path.findLastAnyOf(setOf("miktex\\bin"))?.first ?: (path.length - 1)
-                results.add(path.substring(0, index))
+                if (index > 0) {
+                    results.add(path.substring(0, index))
+                }
             }
         }
         else {

--- a/src/nl/hannahsten/texifyidea/util/General.kt
+++ b/src/nl/hannahsten/texifyidea/util/General.kt
@@ -12,6 +12,10 @@ import java.util.regex.Pattern
 val Boolean.int: Int
     get() = if (this) 1 else 0
 
+// Copied from grazie utils
+fun Boolean?.orTrue() = this ?: true
+fun Boolean?.orFalse() = this ?: false
+
 /**
  * Creates a pair of two objects, analogous to [to].
  */

